File: typescript/packages/bench/src/repositories/customerRepository.ts
/** Read side over seeded customers. */

import type { Repository } from "@atelier/core";
import type { Customer } from "@atelier/core";

/** Read side over seeded customers. */
export class CustomerRepository implements Repository<number, Customer> {
    constructor(private readonly records: readonly Customer[] = []) {}

    find(id: number): Customer | undefined {
        return this.records.find((record) => record.id === id);
    }

    all(): readonly Customer[] {
        return this.records;
    }

    count(): number {
        return this.records.length;
    }

    /** Customers reachable by phone or email. */
    reachable(): readonly Customer[] {
        return this.records.filter((customer) => customer.isReachable());
    }
}

File: rust/crates/atelier-bench/src/repositories/customer_repository.rs
//! Read side over seeded customers.

use atelier_core::Customer;
use atelier_core::contracts::repository::Repository;

/// Read side over seeded customers.
#[derive(Debug, Clone, Default)]
pub struct CustomerRepository {
    records: Vec<Customer>,
}

impl CustomerRepository {
    pub fn new(records: Vec<Customer>) -> Self {
        Self { records }
    }

    /// Customers reachable by phone or email.
    pub fn reachable(&self) -> Vec<&Customer> {
        self.records
            .iter()
            .filter(|customer| customer.is_reachable())
            .collect()
    }
}

impl Repository for CustomerRepository {
    type Id = u32;
    type Record = Customer;

    fn find(&self, id: Self::Id) -> Option<&Self::Record> {
        self.records.iter().find(|record| record.id == id)
    }

    fn all(&self) -> &[Self::Record] {
        &self.records
    }
}
